src/source/Crypto/Dtls_mbedtls.c

```diff
@@ -0,0 +1,739 @@
+    if (certificateBits == 0) {
+        certificateBits = GENERATED_CERTIFICATE_BITS;
+    }
+
+    if (certCount == 0) {
+        CHK_STATUS(createCertificateAndKey(
+              certificateBits, 
+              generateRSACertificate, 
+              &pDtlsSession->certificates[0].cert,
+              &pDtlsSession->certificates[0].privateKey
+        ));
+        pDtlsSession->certificateCount = 1;
+    } else {
+        for (i = 0; i < certCount; i++) {
+            CHK_STATUS(copyCertificateAndKey(
+                  (mbedtls_x509_crt*) pRtcCertificates[i].pCertificate,
+                  (mbedtls_pk_context*) pRtcCertificates[i].pPrivateKey,
+                  &pDtlsSession->certificates[i]
+            ));
+            // in case of a failure in between, we'll only free up to current position
+            pDtlsSession->certificateCount++;
+        }
+    }
+
+    // Generate and store the certificate fingerprints
+    for (i = 0; i < pDtlsSession->certificateCount; i++) {
+        pCertInfo = pDtlsSession->certificates + i;
+        CHK_STATUS(dtlsCertificateFingerprint(&pCertInfo->cert, pCertInfo->fingerprint));
+    }
+    *ppDtlsSession = pDtlsSession;
+
+CleanUp:
+
+    CHK_LOG_ERR(retStatus);
+
+    if (STATUS_FAILED(retStatus) && pDtlsSession != NULL) {
+        freeDtlsSession(&pDtlsSession);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS freeDtlsSession(PDtlsSession *ppDtlsSession)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT32 i;
+    PDtlsSessionCertificateInfo pCertInfo;
+    PDtlsSession pDtlsSession;
+
+    CHK(ppDtlsSession != NULL, STATUS_NULL_ARG);
+
+    pDtlsSession = *ppDtlsSession;
+    CHK(pDtlsSession != NULL, retStatus);
+
+    if (pDtlsSession->timerId != UINT32_MAX) {
+        timerQueueCancelTimer(pDtlsSession->timerQueueHandle, pDtlsSession->timerId, (UINT64) pDtlsSession);
+    }
+
+    for (i = 0; i < pDtlsSession->certificateCount; i++) {
+        pCertInfo = pDtlsSession->certificates + i;
+        freeCertificateAndKey(&pCertInfo->cert, &pCertInfo->privateKey);
+    }
+    mbedtls_entropy_free(&pDtlsSession->entropy);
+    mbedtls_ctr_drbg_free(&pDtlsSession->ctrDrbg);
+    mbedtls_ssl_config_free(&pDtlsSession->sslCtxConfig);
+    mbedtls_ssl_free(&pDtlsSession->sslCtx);
+
+    freeIOBuffer(&pDtlsSession->pReadBuffer);
+    if (IS_VALID_MUTEX_VALUE(pDtlsSession->sslLock)) {
+        MUTEX_FREE(pDtlsSession->sslLock);
+    }
+    SAFE_MEMFREE(*ppDtlsSession);
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+INT32 dtlsSessionSendCallback(PVOID customData, const unsigned char *pBuf, ULONG len)
+{
+  STATUS retStatus = STATUS_SUCCESS;
+  PDtlsSession pDtlsSession = (PDtlsSession) customData;
+
+  CHK(pDtlsSession != NULL, STATUS_NULL_ARG);
+
+  pDtlsSession->dtlsSessionCallbacks.outboundPacketFn(pDtlsSession->dtlsSessionCallbacks.outBoundPacketFnCustomData, (PBYTE) pBuf, len);
+
+CleanUp:
+
+  return STATUS_FAILED(retStatus) ? -retStatus : len;
```

Comment:
Do we have any consumers aware of this? Perhaps you should document in a comment or something

Reply:
This is an internal status code. I had to multiply by -1 so that mbedtls thinks that it's an error. When it's an error, `mbedtls_ssl_read` will return an error and reach this block, https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/565/files#diff-c1ebb6c5ff118358fc32fd996dd11f8bR169-R172. The status that we'll return to the customer is going to be `STATUS_INTERNAL_ERROR`.

This is definitely not perfect since the error code can be ambiguous, it can be the same error code from mbedtls. But, at least, we'll still log the error code which might help us in debugging, https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/565/files#diff-c1ebb6c5ff118358fc32fd996dd11f8bR169.